crates/block-producer/src/mempool/state.rs

```diff
@@ -206,53 +209,93 @@ impl AccountUpdates {
         self.to
             .keys()
             .find(|commitment| !self.from.contains_key(commitment))
+            .or(self.pass_through.keys().next())
             .copied()
             .unwrap_or_default()
     }
 
     fn is_empty(&self) -> bool {
-        self.from.is_empty() && self.to.is_empty()
+        self.from.is_empty() && self.to.is_empty() && self.pass_through.is_empty()
     }
 
-    fn remove(&mut self, from: Word, to: Word) {
-        let from_removed = self
-            .from
-            .remove(&from)
-            .expect("should only be removing account updates from nodes that are present");
-        let to_removed = self
-            .to
-            .remove(&to)
-            .expect("should only be removing account updates from nodes that are present");
-        assert_eq!(
-            from_removed, to_removed,
-            "Account updates should be removed as a pair with the same node ID"
-        );
+    fn remove(&mut self, id: NodeId, from: Word, to: Word) {
+        if from == to {
+            let entry = self.pass_through.entry(from).or_default();
+            assert!(
+                entry.remove(&id),
+                "Account pass through commitment removal of {from} for {id:?} does not exist"
+            );
+            if entry.is_empty() {
+                self.pass_through.remove(&from);
+            }
+        } else {
+            let from_removed = self
+                .from
+                .remove(&from)
+                .expect("should only be removing account updates from nodes that are present");
+            let to_removed = self
+                .to
+                .remove(&to)
+                .expect("should only be removing account updates from nodes that are present");
+            assert_eq!(
+                from_removed, to_removed,
+                "Account updates should be removed as a pair with the same node ID"
+            );
+            assert_eq!(from_removed, id, "Account update removal should match the input node ID",);
+        }
     }
 
     fn insert(&mut self, id: NodeId, from: Word, to: Word) {
-        assert!(
-            self.from.insert(from, id).is_none(),
-            "Account already contained the commitment {from} when inserting {id:?}"
-        );
-        assert!(
-            self.to.insert(to, id).is_none(),
-            "Account already contained the commitment {to} when inserting {id:?}"
-        );
+        if from == to {
+            assert!(
+                self.pass_through.entry(from).or_default().insert(id),
+                "Account already contained the pass through commitment {from} for node {id:?}"
+            );
```

Comment:
Does this mean the node would now panic in a similar scenario to the one before, only now it would require two passthrough transactions in the same batch? Since they will be seen as independent (afaik) two transactions they can end up in the same batch and the updates could trigger this panic

Reply:
Ah. You're right, I think I assumed that account updates for a batch would get merged. This should be true for `ProvenBatch`, but my `ProposedBatch` impl is lazy and doesn't aggregate them.

Thanks, I'll probably fix the proposed batch so that it does aggregation.